### PR TITLE
Refactor/types

### PR DESCRIPTION
**Pull request by Princesseuh**
## Changes

Back in 1997, when Astro was just getting started, someone created an innocent `@types/astro.ts` file, mostly intended for small types used throughout the codebase and types for untyped modules. With time, this file grew with a mix of internal, public, types from dependencies, unused types etc, to a total of 3000 lines of types.

This PR attempt to untangle that file, only keeping public what should actually be public and going away from the easily confused with DefinitelyTyped's stuff `@types` to a normal folder called.. `types`. I also took this opportunity to split the types into multiple files, just so we don't have to scroll 3000 lines of types to find something. No big thoughts went into the splits, just put things where it seemed they'd belong.

Throughout this journey, I discovered that more types than I thought were exposed as part of the public API, perhaps as a mistake.

## Testing

It builds! We don't have a lot of tests regarding types, but the little we do have do pass too. Smoke tests will fail because of docs docgen script, which need to get adapted to this new structure.

## Docs

https://github.com/withastro/docs/pull/9109

### CODE REVIEW

File: packages/astro/src/content/loaders/glob.ts [hunk 1]
```diff
@@ -131,7 +131,7 @@ export function glob(globOptions: GlobOptions): Loader {
 				if (entryType.getRenderFunction) {
 					let render = renderFunctionByContentType.get(entryType);
 					if (!render) {
-						render = await entryType.getRenderFunction(settings);
+						render = await entryType.getRenderFunction(settings.config);
```
Comment (Princesseuh) [inline packages/astro/src/content/loaders/glob.ts:134]: This is the only runtime change I did, the way it was setup, it was exposing `AstroSettings` publicly, which isn't in line with what we typically do

File: packages/astro/client.d.ts [hunk 2]
```diff
@@ -2,12 +2,6 @@
 /// <reference path="./types/content.d.ts" />
 /// <reference path="./types/actions.d.ts" />
 
-// eslint-disable-next-line  @typescript-eslint/no-namespace
-declare namespace App {
-	// eslint-disable-next-line  @typescript-eslint/no-empty-interface
-	export interface Locals {}
-}
-
```
Comment (Princesseuh) [inline packages/astro/client.d.ts:10]: This is unnecessary, the default definition is already that. However, it was done in a `.d.ts` file, which don't get sent to `dist` so it never worked.

File: packages/astro/client.d.ts [hunk 3]
```diff
@@ -191,7 +185,6 @@ declare module '*.md' {
 		file,
 		url,
 		getHeadings,
-		getHeaders,
```
Comment (Princesseuh) [inline packages/astro/client.d.ts:194]: That method hasn't existed in a while.

File: packages/astro/src/jsx/renderer.ts [hunk 4]
```diff
@@ -1,11 +1,8 @@
-import type { AstroRenderer } from '../@types/astro.js';
-import { jsxTransformOptions } from './transform-options.js';
+import type { AstroRenderer } from '../types/public/integrations.js';
 
 const renderer: AstroRenderer = {
 	name: 'astro:jsx',
 	serverEntrypoint: 'astro/jsx/server.js',
-	jsxImportSource: 'astro',
-	jsxTransformOptions,
```
Comment (Princesseuh) [inline packages/astro/src/jsx/renderer.ts:8]: Those options don't seem to exist anymore? There was a deprecated message on both of them.

File: packages/astro/src/runtime/server/endpoint.ts [hunk 5]
```diff
@@ -1,13 +1,16 @@
...
 import { AstroError } from '../../core/errors/errors.js';
 import type { Logger } from '../../core/logger/core.js';
+import type { APIRoute } from '../../types/public/common.js';
+import type { APIContext } from '../../types/public/context.js';
 
 /** Renders an endpoint request to completion, returning the body. */
 export async function renderEndpoint(
-	mod: EndpointHandler,
+	mod: {
+		[method: string]: APIRoute;
```
Comment (Princesseuh) [inline packages/astro/src/runtime/server/endpoint.ts:12]: Type was fully internal to this line, figured there was no point to it being a separate type.

File: packages/astro/src/integrations/hooks.ts [hunk 6]
```diff
@@ -198,10 +197,6 @@ export async function runHookConfigSetup({
 				addWatchFile: (path) => {
 					updatedSettings.watchFiles.push(path instanceof URL ? fileURLToPath(path) : path);
 				},
-				addDevOverlayPlugin: (entrypoint) => {
-					// TODO add a deprecation warning in Astro 5.
```
Comment (Princesseuh) [inline packages/astro/src/integrations/hooks.ts:202]: The TODO here is wrong, this API was unsupported and undocumented before the toolbar actually released, Astro 5 is prime time for fully removing it (the only app that uses it, Sentry, only uses it as a fallback)

**Review by bluwy — APPROVED** [review]
I hope you used some sort of refactoring tool that made this journey less painful 😅 But anyways I stepped through each of them and it looks fine to me.

1. Should we also update https://github.com/withastro/astro/blob/5e928f5948a2d48e5a3a88bc8f91a56b09c2169f/CONTRIBUTING.md#L292
2. In what case would we use the ts files in `types/*.ts` (not the public folder)? Should we place private shared types in `types/astro.ts`, or should we lean towards putting types inside each domains' file/folder instead? (for example, there's a lot of `types.ts` file.

**Comment by Princesseuh** [comment]
Yea, should definitely update that!

I'm not sure! I think that it's fine for some very specific types that are used everywhere perhaps, but overall it seems like over time we've done more and more random `types` files closer to where the feature live. Do you have a preference?

**Comment by bluwy** [comment]
I like to group them by domain, so the `types.ts` file is fine for me, or if it's inlined to the file that uses it first. If we're ok with this pattern too, it would be nice to put a comment in `types/astro.ts` that describes when to use or not use it. Maybe for types that really encompasses Astro, like `AstroSettings` is fine to be there though.

**Comment by Princesseuh** [comment]
I'll add some documentation all around

**Review by ematipico — APPROVED** [review]

**Pull request merged**